asset-transfer-private-data/chaincode-java/src/test/java/org/hyperledger/fabric/samples/privatedata/AssetTransferTest.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.hyperledger.fabric.samples.privatedata;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.ThrowableAssert.catchThrowable;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.hyperledger.fabric.samples.privatedata.AssetTransfer.AGREEMENT_KEYPREFIX;
+import static org.hyperledger.fabric.samples.privatedata.AssetTransfer.ASSET_COLLECTION_NAME;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyZeroInteractions;
+import static org.mockito.Mockito.when;
+
+import java.io.IOException;
+import java.security.cert.CertificateException;
+import java.util.HashMap;
+import java.util.Map;
+import org.hyperledger.fabric.contract.ClientIdentity;
+import org.hyperledger.fabric.contract.Context;
+import org.hyperledger.fabric.shim.ChaincodeException;
+import org.hyperledger.fabric.shim.ChaincodeStub;
+import org.hyperledger.fabric.shim.ledger.CompositeKey;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+public final class AssetTransferTest {
```

Comment:
This test is good, although I don't think we necessarily need it for every sample.
I would rather see an integration test using the existing javascript application in ci/azure-pipelines.yml , that would ensure everything works end-to-end and that there is functional parity between the go chaincode and java chaincode. Not to mention it would be less work to re-use the existing application to drive the end-to-end test.

Reply:
Agree with JS-app & CI-pipeline to drive auto integration test.

 I still think few good/complex chaincode unit test samples, goes a long way for chaincode developer experience.
Unit tests (especially) in chaincode saves a lot of dev time. This test shows some advanced ClientIdentity & Context mocking in java. If we have further discussion here on this and find this useful, I can extend the PrivData tutorial to describe the unit-test.

My personal experience was chaincode dev iteration takes longer than normal development, if you have to deploy to a network and test with some kind of app.

Reply:
Sounds good! I agree it would be good to mention the unit test approach in the https://hyperledger-fabric.readthedocs.io/en/latest/chaincode4ade.html tutorial as a baseline, and then mention any of the more advanced approaches in the other advanced tutorials.